command/resource/client/grpc-client.go

```diff
@@ -31,9 +32,32 @@ func NewGRPCClient(config *GRPCConfig) (*GRPCClient, error) {
 }
 
 func dial(c *GRPCConfig) (*grpc.ClientConn, error) {
-	// TODO: decide if we use TLS mode based on the config
-	dialOpts := []grpc.DialOption{
-		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	err := checkCertificates(c)
+	if err != nil {
+		return nil, err
+	}
+	var dialOpts = []grpc.DialOption{}
+	if c.GRPCTLS {
+		tlsConfig, err := SetupTLSConfig(c)
+		if err != nil {
+			return nil, fmt.Errorf("failed to setup tls config when tried to establish grpc call: %w", err)
+		}
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+	} else {
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(insecure.NewCredentials()))
 	}
+
 	return grpc.Dial(c.Address, dialOpts...)
 }
+
+func checkCertificates(c *GRPCConfig) error {
+	if c.GRPCTLS {
+		if (c.CAFile != "" || c.CAPath != "") && (c.CertFile != "" && c.KeyFile != "") {
```

Comment:
I don't think this is quite right.

Client cert authentication will not typically be used for gRPC but instead only used for the internal_rpc multiplexed listener. Therefore the cert file and key file configurations should be optional. 

We should allow the following scenarios:

| GRPC TLS Enabled | CA File/Path Provided | Client Key Provided | Client Cert Provided | Error |
| ---------------- | --------------------- | ------------------- | -------------------------- | ----- |
| Yes | Yes | Yes | Yes | No |
| Yes | Yes | No  | No  | No |
| Yes | Yes | Yes | No  | Yes (must provide cert when key is provided) |
| Yes | Yes | No  | Yes | Yes (must provide key when client cert is provided) |
| Yes | No  | Yes | Yes | No (go-rootcerts will use the systems default CAs) |
| Yes | No  | Yes | No  | Yes (must provide cert when key is provided) |
| Yes | No  | No  | Yes | Yes (must provide key when cert is provided) |
| No  | *   | *   | *   | No |

Additionally we should check if SSL verification should be enabled (previously we used the `CONSUL_HTTP_SSL_VERIFY` env var for this. When not enabled we should set the corresponding `InsecureSkipVerify` flag within the `tls.Config` to disable verification.

Reply:
Hi @mkeeler 
Thanks for making this chart, it helps a lot!

1. In order to make the check function easier to review, I just elaborate my changes in _verbose_ mode. Given GRPC TLS Enabled, we have 8 situations(3 variables so 2^3, I discard the ssl disabled situation):


| GRPC TLS Enabled | CA File/Path Provided | Client Key Provided | Client Cert Provided | Error |
| ---------------- | --------------------- | ------------------- | -------------------------- | ----- |
| Yes | Yes | Yes | Yes | No |
| Yes | Yes | No  | No  | No |
| Yes | Yes | Yes | No  | Yes (must provide cert when key is provided) |
| Yes | Yes | No  | Yes | Yes (must provide key when client cert is provided) |
| Yes | No  | Yes | Yes | No (go-rootcerts will use the systems default CAs) |
| Yes | No  | Yes | No  | Yes (must provide cert when key is provided) |
| Yes | No  | No  | Yes | Yes (must provide key when cert is provided) |
| Yes  | No   | No   | No   | No(go-rootcerts will use the systems default CAs) |

This makes things easier to check as we can see it only errors when the user only presents one of client cert and client key, which makes sense: client is not required to self-prove and if the client missed to present CA, we make it up by pulling the default ones.

2. I added the variable called `CONSUL_GRPC_SSL_VERIFY` to do the similar work as `CONSUL_HTTP_SSL_VERIFY`